alibi_detect/utils/pytorch/kernels.py

```diff
@@ -20,21 +56,307 @@ def sigma_median(x: torch.Tensor, y: torch.Tensor, dist: torch.Tensor) -> torch.
     n = n if (x[:n] == y[:n]).all() and x.shape == y.shape else 0
     n_median = n + (np.prod(dist.shape) - n) // 2 - 1
     sigma = (.5 * dist.flatten().sort().values[n_median].unsqueeze(dim=-1)) ** .5
-    return sigma
+    return sigma.log()
 
 
-class GaussianRBF(nn.Module):
+class KernelParameter:
     def __init__(
         self,
-        sigma: Optional[torch.Tensor] = None,
-        init_sigma_fn: Callable = sigma_median,
-        trainable: bool = False
+        value: torch.Tensor = None,
+        init_fn: Optional[Callable] = None,
+        requires_grad: bool = False,
+        requires_init: bool = False
+    ) -> None:
+        """
+        Parameter class for kernels.
+
+        Parameters
+        ----------
+        value
+            The pre-specified value of the parameter.
+        init_fn
+            The function used to initialize the parameter.
+        requires_grad
+            Whether the parameter requires gradient.
+        requires_init
+            Whether the parameter requires initialization.
+        """
+        super().__init__()
+        self.value = nn.Parameter(value if value is not None else torch.ones(1),
+                                  requires_grad=requires_grad)
+        self.init_fn = init_fn
+        self.requires_init = requires_init
+
+
+class BaseKernel(nn.Module):
+    def __init__(self, active_dims: list = None, feature_axis: int = -1) -> None:
+        """
+        The base class for all kernels.
+
+        Parameters
+        ----------
+        active_dims
+            Indices of the dimensions of the feature to be used for the kernel. If None, all dimensions are used.
+        feature_axis
+            Axis of the feature dimension.
+        """
+        super().__init__()
+        self.parameter_dict: dict = {}
+        if active_dims is not None:
+            self.active_dims = torch.as_tensor(active_dims)
+        else:
+            self.active_dims = None
+        self.feature_axis = feature_axis
+        self.init_required = False
+
+    @abstractmethod
+    def kernel_function(self, x: Union[np.ndarray, torch.Tensor], y: Union[np.ndarray, torch.Tensor],
+                        infer_parameter: Optional[bool] = False) -> torch.Tensor:
+        raise NotImplementedError
+
+    def forward(self, x: Union[np.ndarray, torch.Tensor], y: Union[np.ndarray, torch.Tensor],
+                infer_parameter: bool = False) -> torch.Tensor:
+        x, y = torch.as_tensor(x), torch.as_tensor(y)
+        if self.active_dims is not None:
+            x = torch.index_select(x, self.feature_axis, self.active_dims)
+            y = torch.index_select(y, self.feature_axis, self.active_dims)
+        if len(self.parameter_dict) > 0:
+            return self.kernel_function(x, y, infer_parameter)
+        else:
+            return self.kernel_function(x, y)
+
+    def __add__(
+        self,
+        other: Union['BaseKernel', 'SumKernel', 'ProductKernel', torch.Tensor]
```

Comment:
Same comment as at the top of this file: any reason why `"BaseKernel"` and not just `BaseKernel` is used as type? Silly question maybe, so just want to know :) @jklaise ?

Reply:
Personally don't like `other` as a argument name, especially given absence of docstrings it is not easy to understand the logic straight away. It would be helpful to have a more clear name.

Reply:
I think `other` is a standard name for implementing `__add__` (and other (!) Python data model methods), though if a different name would make the code clearer then I don't see a reason not to do it (perhaps the code needs some comments anyway, explaining the handling of sum kernels?).

Reply:
Ok, got it, didn't know that. Clarifying comments would be helpful in that case.